src/python/pants/rules/core/test.py

```diff
@@ -0,0 +1,50 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from builtins import str
+
+from pants.base.exceptions import GracefulTerminationException
+from pants.engine.console import Console
+from pants.engine.legacy.graph import HydratedTarget, HydratedTargets
+from pants.engine.rules import console_rule, rule
+from pants.engine.selectors import Get, Select
+from pants.rules.core.core_test_model import GenericTestResult, Status
+from pants.rules.core.python_test_runner import PyTestResult
+
+
+@console_rule('test', [Select(Console), Select(HydratedTargets)])
+def fast_test(console, targets):
+  test_results = yield [Get(GenericTestResult, HydratedTarget, target) for target in targets]
```

Comment:
So, I believe that further decoupling is possible here.

It should be possible for this rule to `Select(BuildFileAddresses)` (which is the first stage of parsing: "addresses for the spec roots"), and then select a TestResult for each of those. That should allow build file parsing to operate in parallel with test runs.

Reply:
Done